### DISCUSSION TITLE
loadAuthToken middleware could coninue by checking if the e.Auth already exists.

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#5800] [Q&A]
If this was added to loadAuthToken middleware then other custom middlewares could be added in front of it by taking advantage of the new `Priority` field.

```golang
			if e.Auth != nil {
				return e.Next()
			}
```

This wouldn't harm default Pocketbase at all, while greatly enhancing the JS and Golang frameworks.

```golang
func loadAuthToken() *hook.Handler[*core.RequestEvent] {
	return &hook.Handler[*core.RequestEvent]{
		Id:       DefaultLoadAuthTokenMiddlewareId,
		Priority: DefaultLoadAuthTokenMiddlewarePriority,
		Func: func(e *core.RequestEvent) error {
			if e.Auth != nil {
				return e.Next()
			}

			token := getAuthTokenFromRequest(e)
			if token == "" {
				return e.Next()
			}

			record, err := e.App.FindAuthRecordByToken(token, core.TokenTypeAuth)
			if err != nil {
				e.App.Logger().Debug("loadAuthToken failure", "error", err)
			} else if record != nil {
				e.Auth = record
			}

			return e.Next()
		},
	}
}
```

I'd like to see something like this included in Pocketbase for framework users.

```golang
package middleware

import (
	"github.com/pocketbase/pocketbase/apis"
	"github.com/pocketbase/pocketbase/core"
	"github.com/pocketbase/pocketbase/tools/hook"
)

const LoadAuthCookieDeferredMiddlewarePriority = apis.DefaultLoadAuthTokenMiddlewarePriority + 1
const LoadAuthCookiePreemptMiddlewarePriority = apis.DefaultLoadAuthTokenMiddlewarePriority - 1

func LoadAuthCookie(cookieName string, priority int) *hook.Handler[*core.RequestEvent] {
	return &hook.Handler[*core.RequestEvent]{
		Func: func(e *core.RequestEvent) error {
			if e.Auth != nil {
				return e.Next()
			}

			tokenCookie, err := e.Request.Cookie(cookieName)
			if err != nil {
				return e.Next()
			}

			if m, err := e.App.FindAuthRecordByToken(tokenCookie.Value, core.TokenTypeAuth); err == nil {
				e.Auth = m
			}

			return e.Next()
		},
		Priority: priority,
	}
}
```

**Comment by ganigeorgiev** [comment]
I've pushed a small change in the `develop` branch that skips the default `loadAuthToken` handling if `e.Auth` is already loaded.

**Reply by dschissler** [reply]
Prior to the refactor this equivalent code was a big ugly mess.  Now it seems to be approaching being universally flexible enough that it could be included by default.  I can't imagine what else that it needs.  The tricky and unique parts are going are going to be where the cookie is set in the authentication event hook.

**Reply by ganigeorgiev** [reply]
Cookies and other SSR helpers are for now low priority. Users can create their own server-side "plugins" if they want to. Later in case I need it or there is enough users demand we can always merge/reimplement it in the main repo.

**Reply by dschissler** [reply]
The issue with not having it is that you won't be able to get input about how to further generalize it.  Probably what I provided is not quite parameterized sufficiently to accommodate 80-90% of use cases, but there is no way to find out what else is needed until it's in tree.  I agree that it's low priority though.

The benefit to having more cookie auth handlers and helpers is that it makes it easy to use Pocketbase as a SSR.  IMO just the database ORM landscape is a shitshow in Golang because there is this general advice of "don't even use a framework" that is everywhere on places like Reddit.  (Perhaps Buffalo project tried this but they had to waste too much time creating embed functionality and other things and the project was made up of too many disjoint mini projects).

So then what do you know, once someone made a really good framework (Pocketbase) then it became incredibly popular in a short period of time.  So all of that to say that I think that there is a rather large minority use case of SSR, HTMX or just buildless JS that can leverage Pocketbase's routing and ORM framework in a way that isn't being met by any other project.
